function_calling/src/function_calling.py

```diff
@@ -0,0 +1,198 @@
+        
+      
+    def set_llm(self, api: str):
+        if api == "sambastudio":
+            llm= SambaStudio(
+                model_kwargs={       
+                        "max_tokens_to_generate": 2048,
+                        "select_expert": "Meta-Llama-3-8B-Instruct",
+                        "process_prompt": False
+                    }
+            )
+        elif api == "sambaverse":
+            llm = Sambaverse(
+                sambaverse_model_name="Meta/Meta-Llama-3-70B-Instruct",
+                model_kwargs={       
+                        "max_tokens_to_generate": 2048,
+                        "select_expert": "Meta-Llama-3-70B-Instruct",
+                        "process_prompt": True,
+                        "temperature": 0.01
+                    }
+            )
+        else:
+            raise ValueError(
+                f"Invalid LLM API: {api}, only'sambastudio' and'sambaverse' are supported."
+            )
+        return llm
+    
+      
+        
+    def get_tools_schemas(self, tools: Union[Tool, list] = None, default: Union[Tool, BaseModel]= None):
+        
+        if tools is None:
+            pass
+        elif isinstance(tools, Tool):
+            tools = [tools]
+        tools_schemas = []
+        
+        for tool in tools:
+            tool_schema = tool.get_input_schema().schema()
+            schema = {"name": tool.name, "description": tool_schema["description"], "properties": tool_schema["properties"]}
+            if "required" in schema:
+                schema["required"] = tool_schema["required"]
+            tools_schemas.append(schema)  
+            
+        if default is not None:
+            if isinstance(default, Tool):
+                tool_schema=default.get_input_schema().schema()
+            else:
+                tool_schema = default.schema()
+            schema = {"name": tool_schema["title"], "description": tool_schema["description"], "properties": tool_schema["properties"]}
+            if "required" in schema:
+                schema["required"] = tool_schema["required"]
+            tools_schemas.append(schema)
+                
+        return tools_schemas  
+        
+    def execute(self, tools):
+        tools_map = {tool.name: tool for tool in self.tools}
+        tool_msg = "Tool '{name}'response: {response}"
+        tools_msgs = []
+        if len(tools)==1 and tools[0]["tool"].lower() == "conversationalresponse":
+            final_answer = True
+            return final_answer, tools[0]["tool_input"]["response"]
+        for tool in tools:
+            final_answer = False
+            if tool["tool"].lower()!="conversationalresponse":
+                response = tools_map[tool["tool"].lower()](tool["tool_input"])
+                tools_msgs.append(tool_msg.format(name=tool["tool"], response=str(response)))
+        return final_answer, tools_msgs
+        
+    def jsonFinder(self, input_string):
```

Comment:
Comments (at least a one liner for the function description) and annotations.

Reply:
added